personaelabs/creddd
# [READY FOR REVIEW] Faster prover

## Overview
- Switch to spartan-evm from spartan-ecdsa. spartan-evm not only can be optimistically verified on-chain but also has a faster prover than spartan-ecdsa. This PR introduces the fast prover into creddd. (on-chain verification is still TBD)
- spartan-evm uses a dedicated frontend for writing constraints (spartan-ecdsa uses circom). This is because we initially planned to implement lookups, so a tailored built frontend was necessary. But we don't have lookups yet, so in theory, we can use Circom. But since the frontend of spartan-evm is vertically integrated it's more amenable to wasm compilation and in-browser execution. In particular, circom requires two different files to be downloaded: the wasm witness generator and the circuit binary file which results in > 40MB of download. Compared to this, spartan-evm only needs to download a single 500KB wasm file. The key difference is that saprtan-evm generates the constraints client-side. This constraint generation only takes few hundred milliseconds, so it's a lot faster than downloading and parsing a 40MB circuit file.

## Todos
- [x] Fix server side verification that's failing for some reason https://github.com/personaelabs/spartan-evm/commit/83202388c0911d33e39fe7ee242ab9fc328caff5
- [x] Fix failing wasm instantiation in production build 2e47d15d069ace2264674a61dfd5aacf901dd421

## How to test
- Check that proof generation works and the generated proofs verify.
- Check that existing proofs verify
